### PR TITLE
feat: adaptive window size for tipset syncer

### PR DESCRIPTION
**Pull request by hanabi1224**
## Summary of changes

Changes introduced in this pull request:

- [x] Implement `AdaptiveValueProvider`
- [x] Tests
- [x] Manually test on mainnet with https://forest-snapshots.fra1.cdn.digitaloceanspaces.com/debug/forest_snapshot_mainnet_2023-06-28_height_2988801.car.zst

## Reference issue to close (if applicable)

Closes https://github.com/ChainSafe/forest/issues/3119

## Change checklist

- [x] I have performed a self-review of my own code,
- [x] I have made corresponding changes to the documentation,
- [x] I have added tests that prove my fix is effective or that my feature works (if possible),
- [x] I have made sure the [CHANGELOG][1] is up-to-date. All user-facing changes should be reflected in this document.

[1]: https://github.com/ChainSafe/forest/blob/main/CHANGELOG.md

### CODE REVIEW

File: src/utils/misc/value_provider.rs [hunk 1]
```diff
@@ -0,0 +1,226 @@
...
+                        Some(i)
+                    }
+                })
+                .expect("Infallible");
+        }
+    }
+
+    /// Tracks failure, a value downgrade will be triggered when consecutive failures is greater than `config.downgrade_threshold`
+    pub fn track_failure(&self) {
+        let old = self.consective_failure_counter.fetch_add(1, ORDERING);
```
Comment (hanabi1224) [inline src/utils/misc/value_provider.rs:79]: AtomicUsize does not panic abt overflow so no special handling is needed

**Comment by lemmih** [comment]
I'm not convinced this is the right approach. I don't understand why peers reject requests of 32 blocks, and I'm not sure if we're being a bad network participant by requesting that many blocks. Also, we don't need to request more than the default amount of blocks to fully saturate our pipeline if we use proper streaming logic. The current code is quite awkward and won't request more than 1 batch at a time.

**Comment by lemmih** [comment]
Let's hold off on merging this until we can benchmark against a streaming solution.

**Review by lemmih — CHANGES_REQUESTED** [review]
Blocking until we can benchmark against a streaming solution.

**Pull request closed without merging**